chore: Refactor

peterpaints opened the pull request:
#### What does this PR do?
- Refactors folder structure to follow a general rails convention.
- Refactors code to stay in line with `rubocop`'s style guides.
- Adds a rake task to be run by the heroku scheduler

#### Description of Task to be completed?
N/A

#### Where should the reviewer start?
View file changes on PR

#### How should this be manually tested?
Run `rubocop` while in the apps root directory

#### What are the relevant pivotal tracker stories?
[#159048831](https://www.pivotaltracker.com/story/show/159048831)

Screenshots (if appropriate)
N/A

peterpaints committed:
chore: Update README

- Add heroku url
- Add env setup instructions
- Update startup instructions

peterpaints committed:
chore: Fix rubocop errors

peterpaints committed:
chore: Create rakefile with notifier task

- This is needed for Heroku, since the code in the configure block under routes
  does not run if the Heroku web dyno is idled, which heroku does after 20 mis of unuse

[Delivers #159022097]

peterpaints committed:
chore: Refactor folder structure to follow rails convention

peterpaints committed:
fix: Allow users to see previously selected topics when adding more

- Refactor sidebar into its own view

hmasila commented on Rakefile:
```diff
@@ -0,0 +1,17 @@
+
+desc 'Create digests and send them daily'
+task notifier: :environment do
+  @users = User.all
```
is this variable used outside this context? if not consider using a normal variable instead of instance variable

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+
+post '/register' do
+  validate_params(params)
+  @user = User.first(email: params[:email])
```
no need to put this in a variable if it's only used by the line below. After all it is being redelared in line 10

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+post '/register' do
+  validate_params(params)
+  @user = User.first(email: params[:email])
+  if @user
```
``` ruby
if(User.first(email: params[:email]))
  // do something
end

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+    flash[:danger] = 'User already exists. Please Log In.'
+    redirect '/'
+  else
+    @user = User.new email: params[:email]
```
try to use brackets for methods with parameters eg `User.new(email: params[:email])`

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+    redirect '/'
+  else
+    @user = User.new email: params[:email]
+    @user.password = @user.hash_password params[:password]
```
`hash_password` can be written as a callback in the model. You will have to inlcude the `active_record` gem into your application http://recipes.sinatrarb.com/p/models/active_record

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+    @user.password = @user.hash_password params[:password]
+    @user.save
+
+    if @user.saved?
```
No need to make two calls to the db. `@user.save` returns a boolean that can be used here. So, line 12 and 13 could be something like; 
``` ruby
if @user.save
  // do something
end

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+      session[:email] = @user.email
+      redirect '/topics'
+    else
+      flash[:danger] = 'Could not create user. Please try again.'
```
Since this else part is reused, you can have a method called display_failure

``` ruby
def display_failure(message)
  flash[:danger] = message
  redirect '/
end

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+    redirect '/'
+  end
+
+  if @user.registered_password? params[:password]
```
eern :nitpick - maybe the method name here could be phrased differently eg `@user.authenticate(params[:password])`

hmasila commented on app/controllers/auth.rb:
```diff
@@ -0,0 +1,53 @@
+end
+
+get '/logout' do
+  require_admin
```
use a filter for this and include all routes that reqiure admin

hmasila commented on app/controllers/pages.rb:
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+get '/' do
+  redirect '/dashboard' if session[:email]
```
maybe have a method called `logged_in?` so that this line will read `redirect '/dashboard' if logged_in?`

hmasila commented on app/controllers/pages.rb:
```diff
@@ -0,0 +1,27 @@
+  require_admin
+  @user = User.first(email: session[:email])
+  @digests = @user.userdigests.reverse
+  @digest = @digests[0]
```
we have a ruby method called `first` so `@digests[0]` could be `@digests.first` which IMO sounds more readable

hmasila commented on app/controllers/pages.rb:
```diff
@@ -0,0 +1,27 @@
+end
+
+get '/unsubscribe' do
+  require_admin
```
use filter for this

hmasila commented on app/controllers/pages.rb:
```diff
@@ -0,0 +1,27 @@
+get '/unsubscribe' do
+  require_admin
+  @user = User.first(email: session[:email])
+  @user.unsubscribed = @user.unsubscribed? ? false : true
```
what is the purpose of this line. It looks like a repeatition since `@user.unsubscribed?` returns a boolen too

Sidenote, for methods that return a boolean and you want to assign a boolean, there is no need for ternary operater. The method is self sufficient

hmasila commented on app/controllers/pages.rb:
```diff
@@ -0,0 +1,27 @@
+  @user.unsubscribed = @user.unsubscribed? ? false : true
+  @user.save
+
+  if @user.saved?
```
No need to check persistance here. `@user.save` works just fine

hmasila commented on app/controllers/routes.rb:
```diff
@@ -0,0 +1,40 @@
+require_relative '../models/models'
+require_relative '../helpers/authorization'
+require_relative '../helpers/digests'
+require_relative '../helpers/mailer'
```
maybe require all of them at once `require_relative '../helpers/*`

hmasila commented on app/controllers/routes.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+end
+
+require_relative 'auth'
```
is there a reason why this are required at the end of the file?

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+
+helpers do
+  def res(status_code, message)
+    status status_code
```
Use def with parentheses when there are parameters. Omit the parentheses when the method doesn't accept any parameters. 
https://github.com/rubocop-hq/ruby-style-guide#method-parens

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+end
+
+get '/topics' do
+  require_admin
```
use filters for this

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+get '/topics' do
+  require_admin
+  @topics = Topic.all
+  gon.topic_titles = @topics.map(&:title)
```
what does gon mean?

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+  gon.topic_titles = @topics.map(&:title)
+  @user = User.first(email: session[:email])
+  gon.user_topics = @user.topics.map(&:title) unless @user.topics.empty?
+  @digests = @user.userdigests
```
`userdigests` here seems to be miswritten. If you want to use two words for a method, separate them with an underscore

scondly `@user.userdigests` sounds like a repeatition. If the method `userdigests` is in the user scope, there is no need to have the user prefix. `digests` will just be fine

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+  require_admin
+  @topics = Topic.all
+  gon.topic_titles = @topics.map(&:title)
+  @user = User.first(email: session[:email])
```
this could be a method because it has been repeated severally
Try something like

``` ruby
def current_user(email)
  User.first(email: email)
end
```

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+end
+
+post '/topics' do
+  require_admin
```
use filters for this. I shared in the channel

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+post '/topics' do
+  require_admin
+  content_type :json
+  @user = User.first(email: session[:email])
```
put this in a method that can be re-used

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+end
+
+get '/unfollow/:id' do
+  require_admin
```
use filters for this

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+
+get '/unfollow/:id' do
+  require_admin
+  @user = User.first(email: session[:email])
```
create `current-user` method that can be re-used

hmasila commented on app/controllers/topics.rb:
```diff
@@ -0,0 +1,53 @@
+  @user.topics.delete @topic
+  @user.save
+
+  if @user.saved?
```
use `@user.save` here

hmasila commented on app/helpers/authorization.rb:
```diff
@@ -0,0 +1,35 @@
+
+module Sinatra
+  module Authorization
+    def unauthorized!
```
The `!` symbol is used to represent persistence or defin ite change. For example, it is used when saving to the db, `@user.save!` which means it saves to the db and returns the result or returns validation errors if it fails. It is also used to make persistent changes like `[1,4,5,2,6,9,2].sort!` permanently alters the arrangement of the values.


What's my point? I feel that the use of a bang here is being mis-represented. Maybe you could try something `def unauohorized_page` or `unauthorized_user` or `unauthorized_error_message`

hmasila commented on app/helpers/authorization.rb:
```diff
@@ -0,0 +1,35 @@
+    end
+
+    def authorized?
+      session[:email].nil? ? false : true
```
no need for the ternary. `nil?` is returns a boolean. Actually, all methods with a question mark return a boolen eg `empty?`, `any?`, `even?`, `include?` etc

hmasila commented on app/helpers/digests.rb:
```diff
@@ -0,0 +1,46 @@
+        description: article_json['description'],
+        url: article_json['url']
+      )
+      article if article.save
```
yeah, good use of `save` here

hmasila commented on app/models/article.rb:
```diff
@@ -1,12 +1,12 @@
 
-  belongs_to :userdigest, :required => false
-
+  belongs_to :userdigest, required: false
```
find me to explain how we can enhance this with `belongs_to :user` and `has_many: digests`

hmasila reviewed (changes requested):
Nce work. A few changes here and there requested

peterpaints committed:
fix: Refactor with requested changes

- Add a before filter
- Add hashing callback to user model
- Add email and password validaton to user model
- Use helper method to show failure messages and redirect
- Change Userdigest table to ReadingList
- Avoid instance variables in tests

peterpaints commented:
@hmasila I've made changes as requested. I hope I've been thorough.
